Fix in-toto-sign args (#118) + huge refactor

lukpueh opened the pull request:
This commit started as fix for in-toto/in-toto#118 but ended up as bigger refactor of the signing and signature verification command line tool. IMHO the structure makes more sense now.

It's probably easiest to just take a look at the code as it is now and not think too much about the diff.

The fix:
- fix whitespace problem with `-k KEYS [KEYS ...]` by replacing positional argument `signablepath` with required optional argument `-f FILE` (c.f. [python issue9338](https://bugs.python.org/issue9338) and [stackoverflow](https://stackoverflow.com/questions/26985650/argparse-do-not-catch-positional-arguments-with-nargs))
- fix funny `in-toto-sign sign` idiom by removing subparsers, making signing the default action and adding a `--verify` option

The refactor:
- Reduce redundancy by merging signing functionality
- Declutter huge nested control flow in `main()`
- Reduce docstrings to minimum
- Add early argument checking for conflicting args
- replace `-d, --destination` with `-o, --output`
- replace `-i, --infix` with `-x, --infix`

This PR also updates the unit tests to conform with the interface changes and removes tests for internally called helper functions and tests the interface more thoroughly instead.
It also removes big parts of test case setup as this can be done as part of the tests.

Commit:
Fix in-toto-sign args + huge refactor

This commit started as fix for (in-toto/in-toto#118) but ended
up as bigger refactor of the signing and signature verification
command line tool.

The fix:
 - fix whitespace problem with `-k KEYS [KEYS ...]`
   by replacing positional argument `signablepath` with
   required optional argument `-f FILE`

   c.f. https://bugs.python.org/issue9338,
   https://stackoverflow.com/questions/26985650/argparse-do-not-catch-positional-arguments-with-nargs

 - fix funny `in-toto-sign sign` idiom by removing subparsers, making
   signing the default action and adding a `--verify` option

The refactor:
 - Reduce redundancy by merging signing functionality
 - Declutter huge nested control flow in `main()`
 - Reduce docstrings to minimum
 - Add early argument checking for conflicting args

 - replace `-d, --destination` with `-o, --output`
 - replace `-i, --infix` with `-x, --infix`

Commit:
Update `in-toto-sign` unit tests

Update to conform with interface changes and remove tests for
internally called helper functions and test interface more
thoroughly instead.

Also removes big parts of test case setup as this can be done
as part of the tests.

Commit:
Clarify infix option in header docstring

Commit:
Fix get default params in _load_metadata + test

Fix default parameters when parsing in-toto metadata and add
unittest to trigger parsing error.

Commit:
Add signing tool fail on verify test

Add test to fail signature verification due to general error
(wrong key format).

Commit:
Add unittest for breaking signature

Testing failing verification due to broken signatures was
handled by in-toto-sign tool unittests previously. The recent
changes to the tool removed that specific test.

This commits re-adds such a test to the test_verifylib module,
which is better suited to test this functionality.

awwad commented on in_toto/in_toto_sign.py:
```diff
@@ -18,11 +18,19 @@
+        required to be carried out by a threshold of functionaries.
+        This option is only available for Link metadata.
+        If multiple keys are passed for signing, the short key id of the last
+        key in the arguments list is used as infix.
```
Minor: This is a lot of detail for filename naming explanation, but I don't see where else I'd put it, so perhaps here is the best place for it. Is any explanation required as to the formats of keys or how the signing is done? Is there some place you can refer to that specifies the expected format of key objects or something?

awwad commented:
When a step is expected to be signed off by multiple functionaries, they're generally expected to produce their own, separate link files, which they sign separately. Since one wouldn't expect multiple signatures on a .link file (except in some edge cases, maybe), going out of your way to support that for this tool doesn't make sense, especially if it costs you having to give up one-to-one mapping in the output filenames (using the last key in a set of keys in the filename). So this tool should either not take multiple keys when signing links (but still take multiple keys when signing layouts) or, when provided multiple keys for signing links, produce one link file for each key provided, using that key as the infix in the filename, restoring one-to-one mapping.

With that problem settled, the infixing behavior for filenames should probably be the default (default 
'package.key12345.link' instead of 'package.link'; then you won't need the -x flag, and you'll be less likely to mix files up. If someone wants to produce 'package.link', they can always specify the output filename.

Also, in our chat, we talked about default behavior, and you noted that you thought the default behavior when an output file already exists (replace signature vs add signature) should be consistent for layouts and links, and that the replace functionality should be the default, and that add should not be enabled for links.

Commit:
Update in-toto-sign interface

Following a discussion with @awwad (see gist at
in-toto/in-toto#125) about individual Link metadata files
generally not being signed with more than one key, but rather
having multiple Link metadata files each signed with a single
different key, to fulfill threshold requirements, this commit
makes the following changes:

- Allow only one signing key for Links
- Make "replace signatures" default for Links and Layouts
- Instead of `--replace` provide the option `--append`
  (available for Layouts only, which can have more than
  one signatures)
- Remove `--infix` option and make it default behavior for Links
 (in-toto verification expects links to have a keyid as infix)

awwad commented on in_toto/in_toto_sign.py:
```diff
@@ -16,38 +16,37 @@
 
   # Sign Link and use passed key's short id as filename infix
-  in-toto-sign -f package.c1ae1e51.link -k priv_key -x
+  in-toto-sign -f package.c1ae1e51.link -k priv_key
```
I was briefly confused about -f and -o reading this. Consider:
```
# Sign Link and write to default output filename (using key's short id as infix)
```
or
```
# Sign Link data from specified file. Since -o is not specified, write to
# default output filename, using the short id for priv_key as a filename infix
# (in place of "c1ae1e51")
```

Alternatively, you could add an example that uses -o:

```
# Re-sign Link and write output to a specific filename.
in-toto-sign -f package.old_sig.link -k priv_key -o package.new_sig.link
```

It could also just be obvious enough already. (:

awwad commented on in_toto/in_toto_sign.py:
```diff
@@ -16,38 +16,37 @@
+         "<step name>.<short signing key id>.link"
 
     - verify signatures
 
```
The behavior of in-toto's link signature verification was a little surprising to me (but maybe I've been using TUF too long), so it's probably worth noting here (or perhaps in the _verify_metadata docstring). You probably also want this clarified deeper in, in verifylib.py, and possibly other places:

**When verifying a Link, if there are *additional* signatures on that Link beyond what is required by the Layout, an AuthorizationError("Unauthorized Key! ...") exception is raised, even if all the required signatures are provided.**

(As a side note, it's worth having thought a bit about whether or not this is the ideal behavior, if you haven't. There are integration edge cases where you could imagine wanting extra signatures - supporting outdated Layouts controlled by other distributors, using Link files tangentially for other purposes.)

awwad reviewed (commented):
LGTM otherwise. I haven't carefully reviewed the tests, FWIW.

Commit:
Update in-toto-sign usage example comments

Adopt one of @awwad's suggestions to clarify where link files
are stored when no output parameter is passed.

lukpueh commented:
Thanks for your valuable comments @awwad! I adopted one suggestion (376fae822b14d4333da72c4e16047cffffff2960) and will address the other one with this ticket https://github.com/in-toto/in-toto/issues/135.
